Wenting/move out plotting qc

wenting-x opened the pull request:
Moved out plotting functions and QC functions from class Project, Well and Curve

wenting-x committed:
Add plot.py

wenting-x committed:
Use plotting functions from plot.py

wenting-x committed:
Use plotting functions for Well from plot.py

wenting-x committed:
Use plotting functions for Curve from plot.py

wenting-x committed:
Add unit tests to plotting functions in Project, Well & Curve

wenting-x committed:
Remove unused parameter

wenting-x committed:
Add unit tests for qc functions

wenting-x committed:
Move out qc functions to quality.py

wenting-x committed:
Add unit tests for Curve qc functions

wenting-x committed:
Move out qc functions from Curve

patrick-reinhard commented on welly/curve.py:
```diff
@@ -9,13 +9,16 @@
+from .plot import plot_kde_curve
+from .plot import plot_2d_curve
+from .plot import plot_curve
+from .quality import quality_curve
```
Imports can be grouped per module

wenting-x replied:
I though separating them was the style here.. :D I'll group the imports from plot.py and quality.py modules.

patrick-reinhard commented on welly/quality.py:
```diff
@@ -152,3 +161,210 @@ def spike_locations(curve):
     Return the indicies of the spikes.
     """
     return
+
```
I would put the 'orchestrators' at the top of the module because we are probably going to append more qc tests

patrick-reinhard commented on welly/quality.py:
```diff
@@ -152,3 +161,210 @@ def spike_locations(curve):
+    Run a series of tests and return the corresponding results.
+
+    Args:
+        tests (list): a list of functions.
```
Please add curve in docstring. Also for other orchestrators.

patrick-reinhard reviewed (commented):
Looks good Wenting. See a couple of minor comments

wenting-x committed:
Move qc orchestration functions to top, complete docstring

wenting-x committed:
Group imports

wenting-x committed:
Group imports

fkiraly-shell commented on welly/curve.py:
```diff
@@ -286,92 +285,20 @@ def plot_2d(self, ax=None,
-            return None
-
-    def plot(self, ax=None, legend=None, return_fig=False, alias=None, **kwargs):
+    def plot(self, ax=None, legend=None, return_fig=False, **kwargs):
```
was `alias` used at all here?
If it was, passing the `kwargs` on might cause an error.

wenting-x replied:
`alias` was not used

fkiraly-shell commented on welly/curve.py:
```diff
@@ -286,92 +285,20 @@ def plot_2d(self, ax=None,
         """
-        Plot a curve.
+        Plot a curve. Wrapping plot function from plot.py.
+        By default only show the plot, not return the figure object.
```
Docstring should say what exactly is being plotted.

fkiraly-shell commented on welly/quality.py:
```diff
@@ -10,6 +10,227 @@
+            mnemonics.
+
+    Returns:
+        dict.
```
what does this return exactly?

wenting-x replied:
Added description to the returned object

fkiraly-shell commented on welly/quality.py:
```diff
@@ -10,6 +10,227 @@
+            mnemonics.
+
+    Returns:
+        list. The results. Stick to booleans (True = pass) or ints.
```
how exactly are the results structured?

wenting-x replied:
Added description

fkiraly-shell commented on welly/quality.py:
```diff
@@ -10,6 +10,227 @@
+
+    Args:
+        well (welly.well.Well): Well object.
+        tests (dict): a dictionary of tests, mapping mnemonics to lists of
```
what does test need to contain?

wenting-x replied:
Added description pointing to an example in the tests

fkiraly-shell commented on welly/quality.py:
```diff
@@ -10,6 +10,227 @@
+    Returns:
+        str. An HTML string.
+    """
+    data = well.qc_data(tests, keys=keys, alias=alias)
```
this is calling `qc_data`, but it's actually a function in this module. Remove coupling from `well`.

wenting-x replied:
Done

fkiraly-shell commented on welly/plot.py:
```diff
@@ -0,0 +1,573 @@
+        label (string): What to put on the y-axis. Defaults to curve name.
+        return_fig (bool): If you want to return the MPL figure object. Default True.
+    Returns:
+        None, axis, figure: depending on what you ask for.
```
plot should be described

wenting-x replied:
added description

fkiraly-shell reviewed (changes requested):
I think this is good, in that it moves the QC and plotting concerns in a separate module.
As far as I can see (with exceptions that I noted), the old interface is kept.

Some general comments:

* docstrings should be improved - assumptions on inputs clearer stated (e.g., if it's a `dict` with special structure, describe that structure); if sth is plotted, describe what is plotted
* in the new modules, some functions call back unnecessarily to `curve`/`well` which calls the module again. This coupling should be removed where possible, e.g., other functions within the module should be called instead.
* consider covering more functions with tests - being aware that we may further refactor before wanting to cover all of it
* think about consistent docstring style - perhaps numpydoc? What is the rest of the codebase using?

wenting-x committed:
Improve docstring

wenting-x committed:
Call function from plot.py instead of the function from the Well class

wenting-x committed:
Call functions from quality.py instead of the functions from Well/Curve classes

wenting-x committed:
Add example to tests and alias description

wenting-x committed:
Add description to the returned value

wenting-x committed:
Align docstring style

wenting-x committed:
Fix error in docstring, add description to the returned value

fkiraly-shell commented:
@wenting-x, since you asked me yesterday to review: did you mean @patrick-reinhard's PR (which I just looked at), or also this PR?

You haven't re-requested a review yet and/or commented on the changes or replied to my comments, so I'm not sure.

wenting-x commented:
@fkiraly-shell I have made some small changes to the docstring based on your comments. The review I have asked for is about the changes I have made.

wenting-x committed:
Added a missing parameter, added description to the returned value

wenting-x commented:
> I think this is good, in that it moves the QC and plotting concerns in a separate module. As far as I can see (with exceptions that I noted), the old interface is kept.
> 
> Some general comments:
> 
> * docstrings should be improved - assumptions on inputs clearer stated (e.g., if it's a `dict` with special structure, describe that structure); if sth is plotted, describe what is plotted
> * in the new modules, some functions call back unnecessarily to `curve`/`well` which calls the module again. This coupling should be removed where possible, e.g., other functions within the module should be called instead.
> * consider covering more functions with tests - being aware that we may further refactor before wanting to cover all of it
> * think about consistent docstring style - perhaps numpydoc? What is the rest of the codebase using?

Thanks, Franz! Based on your comments:
* Improved docstring, added descriptions to what a function does or returns.
* Made sure that functions in plot.py or quality.py are not calling the modules.
* In this branch, all the functions we moved out from modules are covered in tests. We will for sure add more tests as we progress.
* The docstring style was kept the same.
